packages/twenty-front/src/modules/object-record/relation-picker/hooks/useAddNewRecordAndOpenRightDrawer.ts

```diff
@@ -0,0 +1,104 @@
+  relationFieldMetadataItem?: FieldMetadataItem;
+  entityId: string;
+};
+export const useAddNewRecordAndOpenRightDrawer = ({
+  relationObjectMetadataNameSingular,
+  relationObjectMetadataItem,
+  relationFieldMetadataItem,
+  entityId,
+}: RecordDetailRelationSectionProps) => {
+  const setViewableRecordId = useSetRecoilState(viewableRecordIdState);
+  const setViewableRecordNameSingular = useSetRecoilState(
+    viewableRecordNameSingularState,
+  );
+
+  const { createOneRecord } = useCreateOneRecord({
+    objectNameSingular: relationObjectMetadataNameSingular,
+  });
+  const { updateOneRecord } = useUpdateOneRecord({
+    objectNameSingular:
+      relationFieldMetadataItem?.relationDefinition?.targetObjectMetadata
+        .nameSingular ?? '',
+  });
+
+  const { openRightDrawer } = useRightDrawer();
+
+  const handleOnCreate =
+    relationObjectMetadataNameSingular !== 'workspaceMember'
+      ? async (searchInput?: string) => {
+          const newRecordId = v4();
+          const labelIdentifierType = getLabelIdentifierFieldMetadataItem(
+            relationObjectMetadataItem,
+          )?.type;
+          const createRecordPayload: {
+            id: string;
+            name:
+              | string
+              | { firstName: string | undefined; lastName: string | undefined };
+            [key: string]: any;
+          } =
+            labelIdentifierType === FieldMetadataType.FullName
+              ? {
+                  id: newRecordId,
+                  name:
+                    searchInput && searchInput.split(' ').length > 1
+                      ? {
+                          firstName: searchInput.split(' ')[0],
+                          lastName: searchInput.split(' ').slice(1).join(' '),
+                        }
+                      : { firstName: searchInput, lastName: '' },
+                }
+              : { id: newRecordId, name: searchInput ?? '' };
+
+          if (
+            relationFieldMetadataItem?.relationDefinition?.direction ===
+            RelationDefinitionType.ManyToOne
+          ) {
+            createRecordPayload[
+              `${relationFieldMetadataItem?.relationDefinition?.targetFieldMetadata.name}Id`
+            ] = entityId;
+          }
+
+          await createOneRecord(createRecordPayload);
+
+          if (
+            relationFieldMetadataItem?.relationDefinition?.direction ===
+            RelationDefinitionType.OneToMany
+          ) {
+            await updateOneRecord({
+              idToUpdate: entityId,
+              updateOneRecordInput: {
+                [`${relationFieldMetadataItem?.relationDefinition?.targetFieldMetadata.name}Id`]:
+                  newRecordId,
+              },
+            });
+          }
+
+          setViewableRecordId(newRecordId);
+          setViewableRecordNameSingular(relationObjectMetadataNameSingular);
+          openRightDrawer(RightDrawerPages.ViewRecord);
+        }
+      : undefined;
```

Comment:
Hard to read with a big ternary like that, prefer early return at the beginning of the function.